### PR TITLE
BUG: Add extra check for failing UTF-8 conversion

### LINKED ISSUE #23809: read_excel crashes python for certain files
```python
import pandas
pandas.read_excel('pandas_crash.xlsx')
```
[pandas_crash.xlsx](https://github.com/pandas-dev/pandas/files/2598556/pandas_crash.xlsx) crashes the Python process.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.6.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en_US
LOCALE: None.None

pandas: 0.23.4
pytest: 3.6.3
pip: 18.1
setuptools: 39.1.0
Cython: 0.28.3
numpy: 1.15.4
scipy: 1.0.1
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.7.2
patsy: None
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: 2.6.9dev0
feather: None
matplotlib: 2.2.2
openpyxl: 2.5.10
xlrd: 1.0.0
xlwt: None
xlsxwriter: None
lxml: 4.1.1
bs4: 4.6.0
html5lib: 0.9999999
sqlalchemy: 1.2.2
pymysql: 0.8.0
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by roberthdevries**
- [x] closes #23809
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/_libs/src/parse_helper.h [hunk 1]
```diff
@@ -34,6 +34,10 @@ int floatify(PyObject *str, double *result, int *maybe_int) {
         data = PyBytes_AS_STRING(str);
     } else if (PyUnicode_Check(str)) {
         tmp = PyUnicode_AsUTF8String(str);
+        if (tmp == NULL) {
+            PyErr_SetString(PyExc_TypeError, "UTF8 decode error");
```
Comment (WillAyd) [inline pandas/_libs/src/parse_helper.h]: Do you need to set the error here? I think the previous call would already set and can just return, no?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Thanks! Can you also check if there are other issues in the tracker this closes? Shouldn't be specific to Excel given where the fix is made

File: pandas/_libs/src/parse_helper.h (hunk 1 above)
Comment (roberthdevries) [reply pandas/_libs/src/parse_helper.h]: Apparently the 'str' object passed is a Unicode type or a subtype. Ref: https://docs.python.org/3/c-api/unicode.html
And it seems to pass that test fine. However, it fails to convert to UTF8, hence the NULL return value.
BTW, I see know that the message is wrong (decode->encode)

Comment (WillAyd) [reply pandas/_libs/src/parse_helper.h]: Let me clarify - so I am asking if you need the PyErr_SetString at all. I think the failing PyUnicode_AsUTF8String call should have already set the global error indicator / message, so I think this is redundant

Comment (roberthdevries) [reply pandas/_libs/src/parse_helper.h]: Removed
I tried to figure out if `PyUnicode_AsUTF8String` sets an exception or not, but I failed to find out.
However a manual test showed an exception so I guess we are covered here.
```
>>> "\udc88".encode()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'utf-8' codec can't encode character '\udc88' in position 0: surrogates not allowed
```

File: pandas/tests/io/excel/test_readers.py [hunk 2]
```diff
@@ -1044,3 +1044,9 @@ def test_excel_read_binary(self, engine, read_ext):
 
         actual = pd.read_excel(data, engine=engine)
         tm.assert_frame_equal(expected, actual)
+
+    def test_excel_high_surrogate(self, read_ext):
+        # GH 23809
+
+        # should not produce a segmentation violation
+        pd.read_excel("high_surrogate.xlsx")
```
Comment (WillAyd) [inline pandas/tests/io/excel/test_readers.py]: Can you create the expected dataframe and use `tm.assert_frame_equal`?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Implementation looks good; just need to improve the test a bit

File: pandas/_libs/src/parse_helper.h (hunk 1 above)
Comment (WillAyd) [reply pandas/_libs/src/parse_helper.h]: Yea that's generally the pattern for the C API functions. Quoting the guide with respect to the global error indicator:

https://docs.python.org/3/c-api/exceptions.html#exception-handling

> Most C API functions don’t clear this on success, but will set it to indicate the cause of the error on failure. Most C API functions also return an error indicator, usually NULL if they are supposed to return a pointer, or -1 if they return an integer (exception: the PyArg_*() functions return 1 for success and 0 for failure).

I'm not sure which ones don't fall into the "Most" category, but good to know this one does

File: pandas/tests/io/excel/test_readers.py (hunk 2 above)
Comment (roberthdevries) [reply pandas/tests/io/excel/test_readers.py]: Added

**Comment by WillAyd** [comment]
Can you check any impact to the CSV reader as well? Based off of where this fix is made I would imagine this impacts that as well as excel, so maybe should add a test / document fix for both

**Comment by jreback** [comment]
lgtm. defer to @WillAyd for testing.

**Review by datapythonista — APPROVED** [review]
lgtm, thanks

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Great thanks @roberthdevries

**Pull request merged**